File: arcade/python/arcade-intro/05_Island of Knowledge/024_minesweeper.py
def minesweeper(matrix):
    
    # initialize our result matrix
    mineCount = [[0 for j in range(len(matrix[i]))] for i in range(len(matrix))]
    
    # iterate through the matrix
    for row in range(len(matrix)):
        for col in range(len(matrix[0])):
            if matrix[row][col]:
                # we found a mine
                sumOneToNeighbors(mineCount, row, col)
    return mineCount                


def sumOneToNeighbors(mineCount, row, col):
    # Top row
    if row-1 >= 0:
        mineCount[row-1][col] += 1
        if col-1 >= 0:
            mineCount[row-1][col-1] += 1
        if col+1 < len(mineCount[row]):    
            mineCount[row-1][col+1] += 1
    
    # Same row
    if col-1 >= 0:
        mineCount[row][col-1] += 1
    if col+1 < len(mineCount[row]):    
        mineCount[row][col+1] += 1
        
    # Bottom row
    if row+1 < len(mineCount):                
        mineCount[row+1][col] += 1
        if col-1 >= 0:
            mineCount[row+1][col-1] += 1
        if col+1 < len(mineCount[row]):    
            mineCount[row+1][col+1] += 1

print(minesweeper([[True, False, False],
          [False, True, False],
          [False, False, False]]))

File: arcade/go/arcade-intro/05_Island of Knowledge/024_minesweeper.go
func minesweeper(matrix [][]bool) [][]int {
	result := [][]int{}

	for row := 0; row < len(matrix); row++ {
		result = append(result, []int{})
		for col := 0; col < len(matrix[row]); col++ {
			result[row] = append(result[row], countMines(matrix, row, col))
		}
	}

	return result
}

func countMines(matrix [][]bool, row int, col int) int {
	count := 0
	for i := row - 1; i < row+2; i++ {
		for j := col - 1; j < col+2; j++ {
			if i >= 0 && i < len(matrix) && j >= 0 && j < len(matrix[row]) {
				if matrix[i][j] && (row != i || col != j) {
					count++
				}
			}
		}
	}
	return count
}
